api/client/client.ts

```diff
@@ -0,0 +1,292 @@
+  // less clear benefits.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  asCspResponse(): CspResponse<any, any> & { name: string } {
+    return {
+      code: 400,
+      name: this.name,
+      content: {
+        request: this.request,
+        response: this.response?.data,
+      },
+    };
+  }
+}
+
+export interface IAtatClient {
+  // Operations defined directly in the specification
+  addPortfolio(
+    request: types.AddPortfolioRequest
+  ): Promise<types.AddPortfolioResponseSync | types.AddPortfolioResponseAsync>;
+  getPortfolioById(request: types.GetPortfolioRequest): Promise<types.GetPortfolioResponse>;
+  patchPortfolio(
+    request: types.PatchPortfolioRequest
+  ): Promise<types.PatchPortfolioResponseSync | types.PatchPortfolioResponseAsync>;
+  getCostsByPortfolio(request: types.GetCostsByPortfolioRequest): Promise<types.GetCostsByPortfolioResponse>;
+  addTaskOrder(
+    request: types.AddTaskOrderRequest
+  ): Promise<types.AddTaskOrderResponseSync | types.AddTaskOrderResponseAsync>;
+  getCostsByClin(request: types.GetCostsByClinRequest): Promise<types.GetCostsByClinResponse>;
+
+  // Operations that check for the status of a previously-issued provisioning request
+  getProvisioningStatus(request: types.GetProvisioningStatusRequest): Promise<types.GetProvisioningStatusResponse>;
+}
+
+/**
+ * An ATAT API specification client for a single Cloud Service Provider.
+ *
+ * This expects to authenicate using an OAuth 2.0 Bearer Token to a single
+ * cloud service provider, using the configuration object provided to the
+ * constructor.
+ */
+export class AtatClient implements IAtatClient {
+  /**
+   * The version of the ATAT CSP API specification that is supported by this version
+   * of the client. This should match the `info.version` from the OpenAPI specification.
+   */
+  public static supportedApiVersion = "v0.3.0";
+
+  private readonly client: AxiosInstance;
+  private readonly logger: ILogger;
+
+  constructor(authToken: string, cspConfiguration: CspConfigurationItem, logger?: ILogger) {
+    this.client = axios.create({
+      baseURL: cspConfiguration.uri,
+      headers: {
+        Authorization: `Bearer ${authToken}`,
+        "User-Agent": "ATAT CSP Api Client",
+        "Content-Type": "application/json",
+        "X-Atat-Api-Version": AtatClient.supportedApiVersion,
+      },
+      // We will perform all validation of the status code within the various API invocation
+      // methods rather than relying on Axios' configuration. This will allow us to return
+      // more custom error information.
+      validateStatus: () => true,
+    });
+
+    // Dynamically convert all values to snake_case to send across the wire, per the API
+    // specification and then convert responses to camelCase for the internal API of this
+    // client.
+    this.client.interceptors.request.use(camelToSnakeRequestInterceptor);
```

Comment:
Nice find to be able to do this in axios rather than our own code.

Reply:
Nevermind, just realized that the interceptor concept is from axios but that we're actually writing the interceptors themselves. Oh well.